### PR TITLE
CB-4681 fix: content downloading in value panel

### PR DESCRIPTION
**Pull request by Wroud**

### CODE REVIEW

File: webapp/packages/core-utils/src/getMIME.ts
Comment (sergeyteleshev) [inline webapp/packages/core-utils/src/getMIME.ts:1]: unit tests are fallen here, please fix

**Comment by sergeyteleshev** [comment]
https://github.com/dbeaver/cloudbeaver/blob/9bc66479a1a13c13b8b69b7e32850c4e4efb9747/webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/isImageValuePresentationAvailable.ts#L16 - add here something like

`return !['application/octet-stream', null].includes(getMIME(value.binary || ''));`

cause we can now see image presentation for Blob which are not images in ValuePanel

File: webapp/packages/core-utils/src/getMIME.ts [hunk 1]
```diff
@@ -6,9 +6,9 @@
  * you may not use this file except in compliance with the License.
  */
 
-export function getMIME(binary: string): string | null {
+export function getMIME(binary: string): string {
   if (binary.length === 0) {
-    return null;
+    return 'application/octet-stream';
```
Comment (sergeyteleshev) [inline webapp/packages/core-utils/src/getMIME.ts:11]: check the conversation tab. this is the reason why image presentation now broken for blobs which are not images

File: webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx [hunk 2]
```diff
@@ -191,59 +181,60 @@ export const ImageValuePresentation: TabContainerPanelComponent<IDataValuePanelP
...
-        notificationService.logException(exception, 'data_viewer_presentation_value_content_download_error');
-      }
-    };
-
-    const valueSize = bytesToSize(isResultSetContentValue(value) ? value.contentLength ?? 0 : 0);
-    const isDownloadable = data.selectedCell && data.contentAction.isDownloadable(data.selectedCell);
+    const valueSize = bytesToSize(isResultSetContentValue(data.cellValue) ? data.cellValue.contentLength ?? 0 : 0);
+    const isTruncatedMessageDisplay = data.truncated && !data.src;
+    const isDownloadable = isTruncatedMessageDisplay && data.selectedCell && data.contentAction.isDownloadable(data.selectedCell);
+    const isCacheDownloading =
```
Comment (sergeyteleshev) [inline webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx]: this part is not quite obvious how it works. may be it is better to move this logic into `ResultSetCacheAction`

so each time when you download the cache we have a ready to go flag here

File: webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx [hunk 3]
```diff
@@ -191,59 +181,60 @@ export const ImageValuePresentation: TabContainerPanelComponent<IDataValuePanelP
...
                 </Button>
               )}
             </QuotaPlaceholder>
-          ) : null}
+          )}
+        </Container>
+        <Container gap dense keepSize>
+          <Container keepSize flexStart center>
+            {data.canSave && (
+              <ActionIconButton title={translate('ui_download')} name="/icons/export.svg" disabled={loading} img onClick={data.download} />
```
Comment (sergeyteleshev) [inline webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx]: please add `debounce` here

**Review by sergeyteleshev — CHANGES_REQUESTED** [review]

File: webapp/packages/core-utils/src/getMIME.ts
Comment (Wroud) [reply webapp/packages/core-utils/src/getMIME.ts:1]: fixed

File: webapp/packages/core-utils/src/getMIME.ts (hunk 1 above)
Comment (Wroud) [reply webapp/packages/core-utils/src/getMIME.ts:11]: fixed

File: webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx (hunk 2 above)
Comment (Wroud) [reply webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx]: fixed

File: webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx (hunk 3 above)
Comment (Wroud) [reply webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx]: fixed

File: webapp/packages/plugin-data-viewer/src/DatabaseDataModel/Actions/ResultSet/ResultSetDataContentAction.ts [hunk 4]
```diff
@@ -165,69 +149,81 @@ export class ResultSetDataContentAction extends DatabaseDataAction<any, IDatabas
```
Comment (sergeyteleshev) [inline webapp/packages/plugin-data-viewer/src/DatabaseDataModel/Actions/ResultSet/ResultSetDataContentAction.ts]: `async getFileDataUrl(element: IResultSetElementKey) {`
lets make this function private. it is not in use outside of this action scope. moreover, it is easy to mess up with logic:

1. use it outside to get link
2. forget to put cached link into the cache so we can use it
3. have a bug somewhere

File: webapp/packages/plugin-data-viewer/src/DatabaseDataModel/Actions/ResultSet/ResultSetDataContentAction.ts [hunk 5]
```diff
@@ -165,69 +149,81 @@ export class ResultSetDataContentAction extends DatabaseDataAction<any, IDatabas
...
 
   async resolveFileDataUrl(element: IResultSetElementKey) {
-    const cachedUrl = this.retrieveFileDataUrlFromCache(element);
+    const cachedUrl = this.retrieveBlobFromCache(element);
 
     if (cachedUrl) {
       return cachedUrl;
     }
 
     const url = await this.getFileDataUrl(element);
```
Comment (sergeyteleshev) [inline webapp/packages/plugin-data-viewer/src/DatabaseDataModel/Actions/ResultSet/ResultSetDataContentAction.ts:149]: this function sets cache loading state as expects
but we have a somekind of freeze in view button cause const `blob = await downloadFromURL(url);` below is not handled with loader, I suppose

I've send you the video

**Review by sergeyteleshev — APPROVED** [review]

File: webapp/packages/plugin-data-viewer/src/DatabaseDataModel/Actions/ResultSet/ResultSetDataContentAction.ts (hunk 4 above)
Comment (Wroud) [reply webapp/packages/plugin-data-viewer/src/DatabaseDataModel/Actions/ResultSet/ResultSetDataContentAction.ts]: fixed

File: webapp/packages/plugin-data-viewer/src/DatabaseDataModel/Actions/ResultSet/ResultSetDataContentAction.ts (hunk 5 above)
Comment (Wroud) [reply webapp/packages/plugin-data-viewer/src/DatabaseDataModel/Actions/ResultSet/ResultSetDataContentAction.ts:149]: will be fixed later

**Pull request merged**